### PR TITLE
implementation of collection caching

### PR DESCRIPTION
**Pull request by saneshark**
Fixes: https://github.com/department-of-veterans-affairs/vets-api/issues/341

### CODE REVIEW

File: lib/rx/client.rb [hunk 1]
```diff
@@ -66,6 +73,11 @@ def post_preferences(params)
 
     private
 
+    def cache_key(action)
+      return nil unless CACHING_ENABLED
```
Comment (markolson) [inline lib/rx/client.rb]: FYA: There's a `FeatureFlipper` class that could use for this _sort_ of toggle, and drive it based of some user info, ENV, Setting, or whatever else.

File: lib/common/models/collection.rb [hunk 2]
```diff
@@ -26,18 +26,66 @@ class Collection
...
       @errors = errors
+      @cache_key = cache_key
       (@data = data) && return if defined?(::WillPaginate::Collection) && data.is_a?(WillPaginate::Collection)
       @data = data.collect do |element|
         element.is_a?(Hash) ? klass.new(element) : element
       end
     end
 
+    def self.redis_namespace
+      @redis_namespace ||= Redis::Namespace.new('Common::Collection', redis: Redis.current)
```
Comment (patrickvinograd) [inline lib/common/models/collection.rb]: More typical convention is to load namespace out of config/redis.yml - I think there's value in having all the namespaces enumerated there.

File: lib/rx/client.rb [hunk 3]
```diff
@@ -13,14 +13,18 @@ class Client < Common::Client::Base
     configuration Rx::Configuration
     client_session Rx::ClientSession
 
+    CACHE_TTL = 3600 * 10 # 10 hour cache
```
Comment (patrickvinograd) [inline lib/rx/client.rb]: Does our ATO allow us to cache Rx data for this length of time?

File: lib/common/models/collection.rb (hunk 2 above)
Comment (patrickvinograd) [inline lib/common/models/collection.rb]: Buuuut I see why that's tricky e.g. with TTL varying per collected class.

Comment (saneshark) [inline lib/common/models/collection.rb]: Yea, I suppose it wouldn't hurt to have defaults specified in redis.yml that this uses, that way redis.yml still draws attention to the fact that this namespace exists. I'll make that change, the key and TTL will still be able to be overridden though on case by case basis.

File: lib/rx/client.rb (hunk 3 above)
Comment (saneshark) [inline lib/rx/client.rb]: No, i'm going to lower this to 1 hour until we can get approval for 10-24 hours.

**Comment by saneshark** [comment]
I think this is the cleanest approach in general, I wouldn't mind switching to memcached for this variety of caching at some later date, but having Common::Collection as a common namespace should keep things sufficiently "namespaced."

**Comment by saneshark** [comment]
@patrickvinograd I added 'common-collection' with a default ttl of 3600 to redis.yml and referencing the constant now. I still won't merge this though until we have ATO confirmation.

File: lib/rx/client.rb [hunk 4]
```diff
@@ -66,6 +73,11 @@ def post_preferences(params)
 
     private
 
+    def cache_key(action)
+      return nil unless config.caching_enabled?
+      "#{session.user_id}:#{action}"
```
Comment (jkassemi) [inline lib/rx/client.rb]: is there or could there ever be a possibility that the `session.user_id would be nil?

File: spec/lib/rx/client_spec.rb [hunk 5]
```diff
@@ -16,69 +16,131 @@
...
-      expect(client_response.body).to eq(status: 'success')
+    it 'gets a single prescription' do
+      VCR.use_cassette('rx_client/prescriptions/gets_a_single_prescription') do
+        expect(client.get_rx(13_650_546)).to be_a(Prescription)
+      end
+    end
+
+    it 'refills a prescription' do
+      VCR.use_cassette('rx_client/prescriptions/refills_a_prescription') do
+        if caching_enabled
```
Comment (jkassemi) [inline spec/lib/rx/client_spec.rb]: did you put this conditional here vs in the `let(:cache_keys)` definition block for any reason in particular? I don't have any feelings about where it goes, just curious.

File: lib/rx/client.rb (hunk 4 above)
Comment (saneshark) [reply lib/rx/client.rb]: user_id in this context is the mhv_correlation_id, so it should never be nil if session is valid, but I can have it return nil if that's the case as precaution.

File: spec/lib/rx/client_spec.rb (hunk 5 above)
Comment (saneshark) [reply spec/lib/rx/client_spec.rb]: i think I just did that to not have to duplicate specs.

**Pull request merged**